### PR TITLE
Fix support search query/result reset functionality

### LINKED ISSUE #43105: (3P) Clearing query in My Home Support Search card causes "no results" UI state
When interacting with the new support search card on Calypso My Home, resetting the search query to empty causes the "no results" state to appear in the UI.

Screen Capture on 2020-06-09 at 15-51-39

#### Steps to reproduce

1. Starting at URL: https://wordpress.com/home/
2. Type `Podcast` into the Support search input field.
3. Wait for results.
4. Clear the search query from the input (using arrow or delete key).
5. See "no results" UI.

#### What I expected

When the search query is reset/removed we should see the default "contextual results" (which is what appears when the component is initially rendered). No "no results" text should appear.

#### What happened instead

We see the "no results" UI state.

#### Browser / OS version

All

#### Context / Source

See also https://github.com/Automattic/wp-calypso/pull/42993

### PR DESCRIPTION
**Pull request by getdave**
This PR fixes an issue in the support search for both:

- the "Get help" card on "My Home"
- the blue "?" FAB inline help

The problem is best summarised in [the Issue](https://github.com/Automattic/wp-calypso/issues/43105):

> When interacting with the new support search card on Calypso My Home, resetting the search query to empty causes the "no results" state to appear in the UI.

## The fix

To fix this I have:

- Fixed to ensure that updating the search _results_ in state also updates the corresponding search _query_ text - otherwise results and query become out of sync!
- ~Add a new action `resetSearchResults` which sets the search results and query to an empty value in the relevant portions of state.~ Not required. Removed.
- Updated implementation of Inline Help (FAB) component to always dispatch the search action even for empty strings. This ensures we get the reset state and contextual results shown. Moreover the components are now almost identical and ripe for the forthcoming refactor!

#### On Master

* Boot Calypso and go to "My Home" on any site.
* See the "Get Help" support search card.
* Make a careful note of the initial state of the UI.
* Type a valid search (eg: `Podcast`). See results.
* Clear the search (via delete or hitting the `X`).
* See invalid state whereby "No results" shows and results are not showing the default "contextual" results. 
* _Repeat_ for the FAB and verify same issue.

#### On this PR

* Boot Calypso and go to "My Home" on any site.
* See the "Get Help" support search card.
* Make a careful note of the initial state of the UI.
* Type a valid search (eg: `Podcast`). See results.
* Clear the search (via delete or hitting the `X`).
* See valid state which should be the same as the initial state of the UI before you interacted with it at all. You should see:
  - there is _no_ "No results" message shown.
  - the search results show the "contextual results".
  - the search field is empty.
* Enter invalid search term and repeat the process above to ensure search state can be reset.
* _Repeat_ for the FAB and verify same fix works here.

## Screenshots

Screen Capture on 2020-06-16 at 12-03-45
Screen Capture on 2020-06-16 at 12-03-06

Fixes https://github.com/Automattic/wp-calypso/issues/43105

### CODE REVIEW

**Review by marekhrabe — APPROVED** [review]
This works well for me - no more unexpected leftover states!

**Comment by getdave** [comment]
Waiting on @retrofox for final approval as he's doing a deep dive on this.

**Comment by retrofox** [comment]
> Waiting on @retrofox for final approval as he's doing a deep dive on this.

Thanks. Sorry, I'm not totally convinced that adding a new action is a good solution. Maybe I'm wrong but I'd like to dive into this. _testing..._

File: client/state/inline-help/reducer.js
Comment (retrofox) [inline client/state/inline-help/reducer.js:82]: I think it will be enough to fix the issue. I think there isn't needed to add extra action.

Comment (getdave) [reply client/state/inline-help/reducer.js:82]: It's a good spot. I will work for the "My Home" version as we always dispatch the _action_ we just don't _track_ the search if there is no query (ie: it's empty):

https://github.com/Automattic/wp-calypso/blob/0f42e3ff06c2c40bc2d425c9bf3c9b9e07ea0b2a/client/my-sites/customer-home/cards/features/help-search/search-card.jsx#L91-L102

However, the FAB won't work as we conditionally dispatch the _action_ only when there is a search query:

https://github.com/Automattic/wp-calypso/blob/0f42e3ff06c2c40bc2d425c9bf3c9b9e07ea0b2a/client/blocks/inline-help/inline-help-search-card.jsx#L72-L86

So let's normalise the inline version so it matches the Home version and we'll be good.

This also shows how we need to extract a common set of components to avoid such issues.

I'll do this tomorrow AM.

Comment (getdave) [reply client/state/inline-help/reducer.js:82]: OK actually I've just fixed it.

**Comment by getdave** [comment]
@retrofox I fixed the issue by normalising. Please can you:

1. Retest it works across both FAB and the My Home. It's easy to miss!
2. Double check the tracking is tracking the right events for both.

Much appreciated.

**Review by retrofox — DISMISSED** [review]
> Double check the tracking is tracking the right events for both.

It removes the _no results_ when the input search is empty. However, it doesn't record search events from the Inline Helper component. :-(

**Comment by retrofox** [comment]
@getdave I've [added a commit](https://github.com/Automattic/wp-calypso/pull/43323/commits/eb2611f9a343fb10216b00bf663d95357ff12fd2) that fixes the bug when tracking events in the inline help component. Please feel free to remove it or test it :-D

**Review by retrofox — COMMENTED** [review]
removing...

File: client/blocks/inline-help/inline-help-search-card.jsx
Comment (obenland) [inline client/blocks/inline-help/inline-help-search-card.jsx:88]: Could we reverse that logic and make it a positive check to make it a little easier to understand what's happening here?

**Review by obenland — APPROVED** [review]
Tested in Calypso live and it fixes the "No results" bug in the My Home search. I couldn't reproduce that happening in the FAB in production.

I think this is good to go once the ternary operator in `requestInlineSearchResultsAndTrack()` was updated

File: client/blocks/inline-help/inline-help-search-card.jsx
Comment (getdave) [reply client/blocks/inline-help/inline-help-search-card.jsx]: Agreed. I've reversed in _both_ locations.

**Comment by getdave** [comment]
@retrofox Thanks for all your work here on sorting out the tracking. I've tested again and it now looks good to me. Could you give it one more run before I deploy pls?

Screen Shot 2020-06-17 at 10 00 05

Screen Shot 2020-06-17 at 09 59 28

File: client/blocks/inline-help/inline-help-search-card.jsx
Comment (retrofox) [inline client/blocks/inline-help/inline-help-search-card.jsx:88]: Debatable! 😆

**Review by retrofox — APPROVED** [review]
LGTM. :shipit:

**Review by retrofox — CHANGES_REQUESTED** [review]
I've found a weird issue in this PR. Let me _Request changes_ before to develop the causes

**Comment by retrofox** [comment]
Sometimes when you type quickly defining many queries the component enters in an infinity loop. I could reproduce it just typing, again and again, paying attention at the network tab of the development console. :-(

https://cloudup.com/caxDw4IXIUu

**Comment by obenland** [comment]
I also noticed longer response times from the API for search terms that don't map to results, up to 30 seconds at times. It doesn't look like it's caused by this PR though as I can reproduce it in production in the FAB itself, too.

**Review by obenland — APPROVED** [review]

**Comment by retrofox** [comment]
> I also noticed longer response times from the API for search terms that don't map to results, up to 30 seconds at times. It doesn't look like it's caused by this PR though as I can reproduce it in production in the FAB itself, too.

This PR introduces this problem for the "Get help" component in the customer home page. I couldn't reproduce it in prod.

**Comment by retrofox** [comment]
> I can reproduce it in production in the FAB itself, too.

I can't. Have you paid attention to the requests and how they increase automatically?

infinity-requests-2

In the meanwhile...
Screen Shot 2020-06-17 at 8 11 11 PM

**Comment by obenland** [comment]
Hm, I didn't notice that before, but I can reproduce that. They do resolve eventually, but the UI doesn't. You're right though, triggering those additional requests seems to be introduced in this PR.

**Comment by getdave** [comment]
@retrofox @obenland I think a regression was introduced when we tried to simplify the PR [here](https://github.com/Automattic/wp-calypso/pull/43323#issuecomment-644782264) when we added the `searchQuery` to the SUCCESS reducer handling.

This caused an infinite loop whereby because the query got updated another request got triggered.

Anyway...I've found a simpler solution and I've [made a PR for that here](https://github.com/Automattic/wp-calypso/pull/43434). It touches fewer files and particularly only 1 component I'd appreciate your review of that instead now.

**Comment by retrofox** [comment]
> Anyway...I've found a simpler solution and I've [made a PR for that here](https://github.com/Automattic/wp-calypso/pull/43434). It touches fewer files and particularly only 1 component I'd appreciate your review of that instead now.

👍 Let's prioritize #43434. Thanks!

**Comment by getdave** [comment]
Closing in favour of https://github.com/Automattic/wp-calypso/pull/43434

**Pull request closed without merging**
